core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -0,0 +1,780 @@
+          versionToGroup.remove(chunk.getObject().getMinorVersion());
+          if (versionToGroup.isEmpty()) {
+            getStateMap(state).remove(rangeKey);
+          }
+        }
+
+        handleRemove(atomicUpdateGroup, RootPartitionRange.of(chunk), chunk.getObject().getMinorVersion(), state);
+        return atomicUpdateGroup;
+      }
+    }
+    return null;
+  }
+
+  private List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> findOvershadowedBy(
+      AtomicUpdateGroup<T> aug,
+      State fromState
+  )
+  {
+    final RootPartitionRange rangeKeyOfGivenAug = RootPartitionRange.of(aug);
+    return findOvershadowedBy(rangeKeyOfGivenAug, aug.getMinorVersion(), fromState);
+  }
+
+  private List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> findOvershadowedBy(
+      RootPartitionRange rangeOfAug,
+      short minorVersion,
+      State fromState
+  )
+  {
+    Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> current = getStateMap(fromState)
+        .floorEntry(rangeOfAug);
+
+    if (current == null) {
+      return Collections.emptyList();
+    }
+
+    // Find the first key for searching for overshadowed atomicUpdateGroup
+    while (true) {
+      final Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> lowerEntry = getStateMap(fromState)
+          .lowerEntry(current.getKey());
+      if (lowerEntry != null && lowerEntry.getKey().startPartitionId == rangeOfAug.startPartitionId) {
+        current = lowerEntry;
+      } else {
+        break;
+      }
+    }
+
+    final List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> found = new ArrayList<>();
+    while (current != null && rangeOfAug.contains(current.getKey())) {
+      final Short2ObjectSortedMap<AtomicUpdateGroup<T>> versionToGroup = current.getValue();
+      found.addAll(versionToGroup.subMap(versionToGroup.firstShortKey(), minorVersion).short2ObjectEntrySet());
+      current = getStateMap(fromState).higherEntry(current.getKey());
+    }
+    return found;
+  }
+
+  /**
+   * Handles addition of the atomicUpdateGroup to the given state
+   */
+  private void handleAdd(AtomicUpdateGroup<T> aug, State newStateOfAug)
+  {
+    if (newStateOfAug == State.STANDBY) {
+      // A standby atomicUpdateGroup becomes visible when its all segments are available.
+      if (aug.isFull()) {
+        // A visible atomicUpdateGroup becomes overshadowed when a fully available standby atomicUpdateGroup becomes
+        // visible which overshadows the current visible one.
+        findOvershadowedBy(aug, State.VISIBLE)
+            .forEach(entry -> transitPartitionChunkState(entry.getValue(), State.VISIBLE, State.OVERSHADOWED));
+        findOvershadowedBy(aug, State.STANDBY)
+            .forEach(entry -> transitPartitionChunkState(entry.getValue(), State.STANDBY, State.OVERSHADOWED));
+        transitPartitionChunkState(aug, State.STANDBY, State.VISIBLE);
+      }
+    }
+  }
+
+  private void addTo(AtomicUpdateGroup<T> aug, State state)
```

Comment:
Suggest differentiating the `handleAdd/addTo/add` method names more.

Maybe something like the following:
`add` -> `addChunk`
`addTo` -> `addAtomicUpdateGroupWithState`
`handleAdd` -> `transitionStandbyGroupWhenFull`

Reply:
Renamed.